data/transactions/logic/eval.go

```diff
@@ -2039,19 +2049,46 @@ func opB(cx *EvalContext) error {
 	return nil
 }
 
+const protoByte = 0xf0
+
 func opCallSub(cx *EvalContext) error {
-	cx.callstack = append(cx.callstack, cx.pc+3)
-	return opB(cx)
+	cx.callstack = append(cx.callstack, frame{
+		retpc:  cx.pc + 3, // retpc is pc _after_ the callsub
+		height: len(cx.stack),
+	})
+	err := opB(cx)
+
+	/* We only set fromCallSub if we know we're jumping to a proto. In opProto,
+	   we confirm we came directly from callsub by checking (and resetting) the
+	   flag. This is really a little handshake between callsub and proto. Done
+	   this way, we don't have to waste time clearing the fromCallsub flag in
+	   every instruction, only in proto since we know we're going there next.
+	*/
+
+	if cx.nextpc < len(cx.program) && cx.program[cx.nextpc] == protoByte {
+		cx.fromCallsub = true
+	}
+	return err
 }
 
 func opRetSub(cx *EvalContext) error {
 	top := len(cx.callstack) - 1
 	if top < 0 {
 		return errors.New("retsub with empty callstack")
 	}
-	target := cx.callstack[top]
+	frame := cx.callstack[top]
```

Comment:
Since all of this processing is in the AVM itself (no Ledger access at all). I don't see how an e2e test can test anything new.  What do you have in mind?

Reply:
An ABI TEAL program that uses the new functionality in a non-trivial way and asserts that we get the expected result. My intuition is that the sum of the new functionality could add up to more than its parts, and so it's worth going beyond unit tests for the "parts".

So technically, this wouldn't really need to be an E2E test, but if we're going to the trouble of creating a such a program, might as well E2E it.

Reply:
I think we agree. The "ends" for this "end to end" test are TEAL source and results of executing AVM.  I speculate that `testAccepts` is sufficient for your needs.  And if it isn't, it ought to be made so that it is. For example, maybe it should allow assertions about something ends up in logs, just as `testPanics` allows an assertion about what error caused the problem.

Reply:
A `testAccepts` variant that also allows asserting logs is a big step forward. But I don't see an easy way `testAccepts` could be modified to provide coverage for the same program over many possible inputs taken as args. Maybe there is another test function that can help here?

Reply:
Make the program a format string and use fmt.Sprintf to try many variants?
`evalCrypto_test.go` does this
```
			src := fmt.Sprintf(source, test.data, hex.EncodeToString(test.r), hex.EncodeToString(s), hex.EncodeToString(x), hex.EncodeToString(y))
			if test.pass {
				testAccepts(t, src, fidoVersion)
			} else {
				testRejects(t, src, fidoVersion)
			}
```

Reply:
If you _really_ want to, you can construct txns, and stuff arguments into ApplicationArgs, then use `testApp` instead, which take an EvalParams containing the txn to execute against (so you'd be able to access ApplicationArgs). But I can't imagine how this functionality has anything to do with whether you get the inputs from AppArgs, or have them built into a new hunk of source code with Sprintf. Therefore I think going full ABI on this is overkill.

Reply:
I was thinking of this sort of example as well. It's probably good enough given that there is very little concern that the arg-passing opcodes would introduce any strange artifacts. It would be more flexible and realistic to allow direct argument handling.

Reply:
Unless we made `frame_dig -1` grab the last ApplicationArg, and so on!  That would vaguely make sense, as a way to get the arguments of the main contract. But I'd be embarrassed to write that spec!